### PR TITLE
Allow specifying custom base version for releases

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Currently automated releases will pull the most recent milestone when creating a new codebase, we'd prefer to be able to specify an exact base version so that we can use a more recent version if one exists, but don't necessarily _have_ to cut a new base version for every release because that add friction to the release process.

## Implementation notes

At compile time, use a template haskell splice to look up the `UNISON_BASE_PATH` env var, and back that into the binary if it's set and parses to a share path. Otherwise fall back to the current logic and try to determine the most recent milestone release from the git tag.

## Test coverage

It works locally but I'll keep an eye on it in the next release as well.

## Open questions

~Do I need to add a 'force rebuild' for `unison-cli` into the release built-path to ensure we don't use a cached version with a stale base version? I think probably yes.~

I added a `stack clean unison-cli` before each 'release' build to address this.

### CODE REVIEW

File: parser-typechecker/src/Unison/Util/Pretty/MegaParsec.hs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+module Unison.Util.Pretty.MegaParsec where
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Pretty/MegaParsec.hs:1]: I moved a few shared combinators here to make the imports a bit more sensible.

Other than that they're unchanged.

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 2]
```diff
@@ -60,6 +65,16 @@ repoPath =
     fmap ReadRemoteNamespaceGit readGitRemoteNamespace
       <|> fmap ReadRemoteNamespaceShare readShareRemoteNamespace
 
+parseReadRemoteNamespace :: String -> String -> Either (P.Pretty P.ColorText) ReadRemoteNamespace
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs:68]: This was moved here from `InputPatterns`

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 3]
```diff
@@ -60,6 +65,16 @@ repoPath =
...
+parseReadRemoteNamespace label input =
+  let printError err = P.lines [P.string "I couldn't parse the repository address given above.", P.prettyPrintParseError input err]
+   in first printError (P.parse repoPath label (Text.pack input))
+
+parseReadShareRemoteNamespace :: String -> String -> Either (P.Pretty P.ColorText) ReadShareRemoteNamespace
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs:73]: This is new, but is just a more specific version of `parseReadRemoteNamespace` which only accepts share paths, not git paths.

File: unison-cli/unison/Main.hs [hunk 4]
```diff
@@ -468,8 +472,14 @@ getConfigFilePath mcodepath = (FP.</> ".unisonConfig") <$> Codebase.getCodebaseD
 
 defaultBaseLib :: Maybe ReadShareRemoteNamespace
 defaultBaseLib =
-  rightMay $
-    runParser VP.defaultBaseLib "version" gitRef
+  let mayBaseSharePath =
+        $( do
```
Comment (ChrisPenner) [inline unison-cli/unison/Main.hs:476]: This is where we look up the env var at compile time.

**Comment by mitchellwrosen** [comment]
> Currently automated releases will pull the most recent milestone when creating a new codebase, we'd prefer to be able to specify an exact base version so that we can use a more recent version if one exists, but don't necessarily have to cut a new base version for every release because that add friction to the release process.

Could you explain this again for me please? If I'm reading right -- which I'm not -- then the release process we already have (of pulling the most recent base version) _is_ the frictionless one.

**Comment by ChrisPenner** [comment]
@mitchellwrosen 

The one we have right now _is_ frictionless,  but it means that we're locked to the most recent MAJOR milestone, so when we release M4d, it will still pull base from `M4`, which means it may be missing names for some new builtins or w/e.

We'd prefer to _allow_ pulling M4d in that case, but not necessarily _require_ that someone cut an M4d release of base before making the release, because sometimes there's nothing new worth making a release for, or maybe Rúnar is just on vacation or something; so that's why we don't want to have it default to _exactly_ the same base version as the release we're cutting, since we may want to have `M4d` just pull `M4` or maybe `M4c` or something.

This PR allows us to specify exactly which version it should pull under every circumstance, so we can just tell the release process "Make the M4d release pull `unison.public.releases.M4d_alpha`" or w/e if we want to 🤷🏼

**Comment by mitchellwrosen** [comment]
> We'd prefer to allow pulling M4d in that case, but not necessarily require that someone cut an M4d release of base before making the release, because sometimes there's nothing new worth making a release for, or maybe Rúnar is just on vacation or something; so that's why we don't want to have it default to exactly the same base version as the release we're cutting, since we may want to have M4d just pull M4 or maybe M4c or something.

I'm not getting something - it seems to me we don't currently require someone cut an `M4d` release of base to go with an `M4d` release of UCM.

**Comment by ChrisPenner** [comment]
> I'm not getting something - it seems to me we don't currently require someone cut an M4d release of base to go with an M4d release of UCM.

Right, but that's because the current logic will always pull `M4`, which is older than folks would like. We should be able to ship `M4d` and have it pull `M4c` if we want it to.

**Comment by mitchellwrosen** [comment]
Oh, ok. Might we want to always pull the latest minor release instead of the latest major release, then?

**Comment by ChrisPenner** [comment]
> Oh, ok. Might we want to always pull the latest minor release instead of the latest major release, then?

Ahh I see where the confusion came from, the issue is that there's no easy mechanism for "detect the latest minor release available", the options are more like "Always pull the current Major release" or "Pull the version which exactly matches the current release version", the latter of which we want to avoid because that version may not have been released on share.

I'm sure we could probably write the code for "get the full list of available versions and figure out which one is best", but that's more work than just specifying the one we want at any given time, which is pretty trivial to do at release time.

**Comment by aryairani** [comment]
> Do I need to add a 'force rebuild' for `unison-cli` into the release built-path to ensure we don't use a cached version with a stale base version? I think probably yes.

This is one of the scary things about using caching in CI, but yes we need to be careful not to build an executable with the wrong contents :)

**Review by aryairani — APPROVED** [review]

**Pull request merged**